components/config/prefs.rs

```diff
@@ -7,251 +7,398 @@ use crate::opts;
 use embedder_traits::resources::{self, Resource};
 use serde_json::{self, Value};
 use std::borrow::ToOwned;
-use std::cmp::max;
 use std::collections::HashMap;
 use std::fs::File;
 use std::io::{stderr, Read, Write};
 use std::path::PathBuf;
-use std::sync::{Arc, RwLock};
+
+use crate::pref_util::Preferences;
+pub use crate::pref_util::{PrefError, PrefValue};
+use gen::Prefs;
 
 lazy_static! {
-    pub static ref PREFS: Preferences = {
-        let defaults = default_prefs();
-        if let Ok(prefs) = read_prefs(&resources::read_string(Resource::Preferences)) {
-            defaults.extend(prefs);
-        }
-        defaults
+    static ref PREFS: Preferences<'static, Prefs> = {
+        let def_prefs: Prefs = serde_json::from_str(&resources::read_string(Resource::Preferences))
+            .expect("Failed to initialize config preferences.");
+        Preferences::new(def_prefs, &gen::PREF_ACCESSORS)
     };
 }
 
-#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
-pub enum PrefValue {
-    Boolean(bool),
-    String(String),
-    Number(f64),
-    Missing,
-}
-
-impl PrefValue {
-    pub fn from_json(data: Value) -> Result<PrefValue, ()> {
-        let value = match data {
-            Value::Bool(x) => PrefValue::Boolean(x),
-            Value::String(x) => PrefValue::String(x),
-            Value::Number(x) => {
-                if let Some(v) = x.as_f64() {
-                    PrefValue::Number(v)
-                } else {
-                    return Err(());
-                }
-            },
-            _ => return Err(()),
-        };
-        Ok(value)
-    }
-
-    pub fn as_boolean(&self) -> Option<bool> {
-        match *self {
-            PrefValue::Boolean(value) => Some(value),
-            _ => None,
-        }
-    }
-
-    pub fn as_string(&self) -> Option<&str> {
-        match *self {
-            PrefValue::String(ref value) => Some(&value),
-            _ => None,
-        }
-    }
-
-    pub fn as_i64(&self) -> Option<i64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as i64),
-            _ => None,
-        }
-    }
-
-    pub fn as_u64(&self) -> Option<u64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as u64),
-            _ => None,
-        }
-    }
-}
-
-#[derive(Clone, Debug, Deserialize, Serialize)]
-pub enum Pref {
-    NoDefault(Arc<PrefValue>),
-    WithDefault(Arc<PrefValue>, Option<Arc<PrefValue>>),
+/// A convenience macro for accessing a preference value using its static path.
+/// Passing an invalid path is a compile-time error.
+#[macro_export]
+macro_rules! get_pref {
```

Comment:
Given reading prefs is much more frequent than setting them, I'd lean about calling this macro just `pref!`, but no big deal in any case.